src/ghost_cell.rs

```diff
@@ -311,23 +311,78 @@ impl<'brand, T: ?Sized> GhostCell<'brand, T> {
+        unsafe {
+            Some((
+                &mut *cell_a.value.get(),
+                &mut *cell_b.value.get(),
+                &mut *cell_c.value.get()
+            ))
+        }
+    }
+}
+
+#[cfg(feature = "experimental-multiple-mutable-borrows")]
+/// Returns `Some(())` if the inputs are distinct, and `None` otherwise.
+fn check_distinct<const N: usize>(arr: [*const (); N]) -> Option<()> {
+    for i in 0..N {
+        for j in i+1..N {
+            if core::ptr::eq(arr[i], arr[j]) {
+                return None;
+            }
+        }
+    }
+    Some(())
+    // TODO: if the array is large enough, sort the values instead.
+}
+
+
+#[cfg(feature = "experimental-multiple-mutable-borrows")]
+macro_rules! generate_multiple_borrow_mut {
+    ( $func_name:ident, $( $input_name:ident, $input_type:ident ),* ) => {
+        pub fn $func_name <'a, 'brand, $( $input_type ,)* >(
+            $( $input_name : &'a GhostCell<'brand, $input_type>, )*
+            _: &'a mut GhostToken<'brand>
+        ) -> Option<( $(&'a mut $input_type,)* )> where T: Sized {
+            // we require that the types are `Sized`, so no fat pointer problems.
+            check_distinct([ $( $input_name as *const (), )* ])?;
             unsafe {
                 Some((
-                    &mut *cell_a.value.get(),
-                    &mut *cell_b.value.get(),
-                    &mut *cell_c.value.get()
+                    $( &mut * $input_name.value.get() ,)*
                 ))
             }
         }
-    }
+    };
+}
+
+#[cfg(feature = "experimental-multiple-mutable-borrows")]
+macro_rules! generate_multiple_borrow_mut_impl {
+    ( $func_name:ident, $first_name:ident, $first_type:ident, $( $input_name:ident, $input_type:ident ),* ) => {
+        impl<'brand, $first_type> GhostCell<'brand, $first_type> {
+            /// doc stub
+            pub fn $func_name <'a, $( $input_type ,)* >(
+                $first_name : &'a GhostCell<'brand, $first_type>,
+                $( $input_name : &'a GhostCell<'brand, $input_type>, )*
+                _: &'a mut GhostToken<'brand>
+            ) -> Option<( &'a mut $first_type, $(&'a mut $input_type,)* )> where $first_type: Sized {
+                // we require that the types are `Sized`, so no fat pointer problems.
+                check_distinct([ $first_name as *const _ as *const(), $( $input_name as *const _ as *const (), )* ])?;
+                unsafe {
+                    Some((
+                        &mut * $first_name.value.get(), $( &mut * $input_name.value.get() ,)*
+                    ))
+                }
+            }
+        }
+    };
 }
 
+#[cfg(feature = "experimental-multiple-mutable-borrows")]
+generate_multiple_borrow_mut_impl!(borrow_mut_four, call_a, T, call_b, Q, cell_c, R, call_d, S);
+#[cfg(feature = "experimental-multiple-mutable-borrows")]
+generate_multiple_borrow_mut_impl!(borrow_mut_five, call_1, T1, call_2, T2, cell_3, T3, call_4, T4, cell_5, T5);
```

Comment:
I'm not a fan of having a function for each arity:

- If you have a function taking a tuple, it's easy enough to use with a non-tuple.
- The reverse, however, is not true.

I'd rather put the difficulty -- handling any arity -- in the base library rather than in the client libraries.